internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// InstallEnvReconciler reconciles a InstallEnv object
+type InstallEnvReconciler struct {
+	client.Client
+	Log                      logrus.FieldLogger
+	Installer                bminventory.InstallerInternals
+	PullSecretUpdatesChannel chan event.GenericEvent
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs/status,verbs=get;update;patch
+
+func (r *InstallEnvReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.ensureISO(ctx, installEnv)
+}
+
+// ensureISO generates ISO for the cluster if needed and will update the condition Reason and Message accordingly.
+// It returns a result that includes ISODownloadURL.
+func (r *InstallEnvReconciler) ensureISO(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var updatedCluster *common.Cluster
+
+	kubeKey := types.NamespacedName{
+		Name:      installEnv.Spec.ClusterRef.Name,
+		Namespace: installEnv.Spec.ClusterRef.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	// Retrieve clusterDeployment
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		// Update that we failed to retrieve the clusterDeployment
+		conditionsv1.SetStatusCondition(&installEnv.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ImageCreatedCondition,
+			Status:  corev1.ConditionUnknown,
+			Reason:  adiiov1alpha1.ImageCreationErrorReason,
+			Message: adiiov1alpha1.ImageStateFailedToCreate + ": " + clusterDeploymentRefErr.Error(),
+		})
+		if updateErr := r.Status().Update(ctx, installEnv); updateErr != nil {
+			r.Log.WithError(updateErr).Error("failed to update installEnv status")
+		}
+		return ctrl.Result{Requeue: false}, nil
```

Comment:
Maybe i wan't clear, you need to check of the error type. `if k8serrors.IsNotFound(err)...`